File: exercise/scala/day24/src/test/scala/doubles/InMemoryToyRepository.scala
package doubles

import domain.core.{Aggregate, Event}
import domain.{Toy, ToyRepository}

import java.util.UUID
import scala.collection.mutable

class InMemoryToyRepository extends ToyRepository {
  private val toys: mutable.Map[UUID, Toy] = mutable.Map.empty
  private val events: mutable.ListBuffer[Event] = mutable.ListBuffer.empty
  
  override def findByName(toyName: String): Option[Toy] =
    toys.values.find(_.getName() == toyName)

  override def findById(id: UUID): Option[Toy] = toys.get(id)

  override def save(aggregate: Toy): Unit = {
    events.clear()
    toys(aggregate.id) = aggregate
    events ++= (aggregate.asInstanceOf[Aggregate].getUncommittedEvents())
    aggregate.asInstanceOf[Aggregate].clearUncommittedEvents()
  }

  def raisedEvents: List[Event] = events.toList
}

File: exercise/kotlin/day24/src/test/kotlin/doubles/InMemoryToyRepository.kt
package doubles

import arrow.core.Option
import arrow.core.toOption
import domain.Toy
import domain.ToyRepository
import domain.core.Aggregate
import domain.core.Event
import java.util.*

class InMemoryToyRepository : ToyRepository {
    private val toys: MutableMap<UUID, Toy> = HashMap()
    private var raisedEvents: MutableList<Event> = mutableListOf()

    override fun findByName(toyName: String): Option<Toy> = toys.values.firstOrNull { it.name() == toyName }.toOption()
    override fun findById(id: UUID): Option<Toy> = toys[id].toOption()

    override fun save(aggregate: Toy) {
        raisedEvents.clear()
        toys[aggregate.id] = aggregate
        raisedEvents.addAll(0, (aggregate as Aggregate).getUncommittedEvents())
        (aggregate as Aggregate).clearUncommittedEvents()
    }

    fun raisedEvents(): List<Event> = raisedEvents
}
